system-addon/content-src/components/TopSites/TopSite.jsx

```diff
@@ -8,59 +8,113 @@ import {
-    imageClassName = "top-site-icon rich-icon";
-    imageStyle = {
-      backgroundColor: link.backgroundColor,
-      backgroundImage: `url(${tippyTopIcon || link.favicon})`
-    };
-  } else {
-    // styles and class names for top sites with screenshot + small icon in top left corner
-    imageClassName = `screenshot${link.screenshot ? " active" : ""}`;
-    imageStyle = {backgroundImage: link.screenshot ? `url(${link.screenshot})` : "none"};
+export class TopSiteLink extends React.PureComponent {
+  constructor(props) {
+    super(props);
+    this.onDragEvent = this.onDragEvent.bind(this);
+  }
 
-    // only show a favicon in top left if it's greater than 16x16
-    if (faviconSize >= MIN_CORNER_FAVICON_SIZE) {
-      showSmallFavicon = true;
-      smallFaviconStyle = {backgroundImage:  `url(${link.favicon})`};
-    } else if (link.screenshot) {
-      // Don't show a small favicon if there is no screenshot, because that
-      // would result in two fallback icons
-      showSmallFavicon = true;
-      smallFaviconFallback = true;
+  /*
+   * Helper to determine whether the drop zone should allow a drop. We only allow
+   * dropping top sites for now.
+   */
+  _allowDrop(e, index) {
+    let draggedIndex = parseInt(e.dataTransfer.getData("text/topsite-index"), 10);
+    if (!isNaN(draggedIndex) && draggedIndex !== index) {
+      return true;
     }
+    return false;
   }
-  return (<li className={topSiteOuterClassName} key={link.guid || link.url}>
-   <a href={link.url} onClick={props.onClick}>
-      <div className="tile" aria-hidden={true} data-fallback={letterFallback}>
-        <div className={imageClassName} style={imageStyle} />
-        {showSmallFavicon && <div
-          className="top-site-icon default-icon"
-          data-fallback={smallFaviconFallback && letterFallback}
-          style={smallFaviconStyle} />}
-     </div>
-     <div className={`title ${link.isPinned ? "pinned" : ""}`}>
-       {link.isPinned && <div className="icon icon-pin-small" />}
-        <span dir="auto">{title}</span>
-     </div>
-   </a>
-   {props.children}
-  </li>);
-};
+  onDragEvent(event) {
+    switch (event.type) {
+      case "dragstart":
+        event.dataTransfer.effectAllowed = "move";
+        event.dataTransfer.setData("text/topsite-index", this.props.index);
+        event.target.blur();
+        this.props.onDragEvent(event, this.props.index, this.props.link, this.props.title);
+        break;
+      case "dragend":
+        this.props.onDragEvent(event);
+        break;
+      case "dragover":
+      case "dragenter":
+      case "dragleave":
+      case "drop":
+        if (this._allowDrop(event, this.props.index)) {
+          event.preventDefault();
+          this.props.onDragEvent(event, this.props.index);
+        }
+        break;
+      default:
+        break;
+    }
+  }
+  render() {
+    const {props} = this;
+    const {link, title, className, isDraggable} = props;
```

Comment:
nit: include `{children, onClick, …} = this.props;` without pulling `props` out the line before